docs/design/design-customized-deviceshifu.md

```diff
@@ -0,0 +1,157 @@
+
+Customizing deviceShifu will enable deviceShifu to handle customized logic such as parsing data from devices.
+
+## Data flow
+
+```mermaid
+sequenceDiagram
+    participant id as IoT device
+    participant cd as customized deviceShifu
+    id->>cd: raw data
+    cd->>cd: parse raw data via handler function
+    cd->>application: parsed data
+```
+Anything between `raw data` and `app-friendly data` are components in `deviceShifu`:
+```mermaid
+  flowchart LR
+    IoT-device --> |raw data|deviceShifu-data-adapter-->deviceShifu-data-handler-->deviceShifu-data-sender-->|app-friendly data|application
+
+```
+
+Examples:
+
+`raw data` provided by IoT device:
+
+```
+root@nginx:/# curl http://deviceshifu-test.deviceshifu.svc.cluster.local/18000851;echo
+```
+
+```json
+{
+    "statusCode": "200",
+    "message":"success",
+    "entity":[{
+        "deviceId":"20990922009",
+        "datatime":"2022-06-30 07:55:51",
+        "eUnit":"℃",
+        "eValue":"37",
+        "eKey":"e3",
+        "eName":"atmosphere temperature",
+        "eNum":"101"
+    },
+    {
+        "deviceId":"20990922009",
+        "datatime":"2022-06-30 07:55:51",
+        "eUnit":"%RH",
+        "eValue":"88",
+        "eKey":"e4",
+        "eName":"atmosphere humidity",
+        "eNum":"102"
+    }]
+}
+```
+
+`parsed data` required by application:
+
+```json
+{
+    [{
+        "code":"20990922009",
+        "name":"atmosphere temperature",
+        "val":"37",
+        "unit":"℃",
+        "exception":"temperature is too high"
+    },
+    {
+        "code":"20990922009",
+        "name":"atmosphere humidity",
+        "val":"88",
+        "unit":"",
+        "exception":"humidity is too high"
+    }]
+}
+```
+
+## Developer workflow
+
+```mermaid
+flowchart LR
+    subgraph devgd[develop general deviceShifu]
+        wdyf[Write deviceShifu yaml files]-->dd[deploy general deviceShifu]
+        dd-->tdc[test device connection]
+    end
+
+    subgraph devcd[develop customized deviceShifu]
+        idhf[implement deviceShifu handler functions]
+        idhf-->tdhf[test deviceShifu handler functions]
+        tdhf-->bcde[build customized deviceShifu executable]
+        bcde-->bcdi[build customized deviceShifu image]
+        bcdi-->dcd[deploy customized deviceShifu]
+        dcd-->tcd[test customized deviceShifu]
+    end
+
+    devgd-->devcd
+```
+
+## Implementation
+
+By default, `deviceShifu` will provide raw data from physical device to the applications.
+
+To make `deviceShifu` able to "translate" the raw data to a more application-friendly format as well as filter out unneeded data, we can provide customized handlers to `deviceShifu`. For example, using Python, we add customized handlers like this:
+
+1. in `deviceshifu/python_customized_handlers/customized_handlers.py`, add a new function with the function name being instruction name.
+2. make sure the function takes a stirng (raw_data) and returns a string (processed_data which app-friendly).
```

Comment:
is the raw data from IoT device? if so what happens if the data is not a string. i.e. binary

Reply:
do we need return an error code if the raw data in case not being parsed or running to some parse exception?
BTW, there is a typo "stirng" in #2